renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -1410,6 +1441,671 @@ struct TestsFailedCallback : public VulkanPixelHistoryCallback
+                                               PixelHistoryShaderCache *shaderCache, uint32_t x,
+                                               uint32_t y, uint32_t sampleMask,
+                                               std::map<uint32_t, rdcarray<int32_t> > events,
+                                               VkQueryPool occlusionPool)
+      : VulkanPixelHistoryCallback(vk, shaderCache, x, y, sampleMask, occlusionPool),
+        m_Events(events)
+  {
+  }
+
+  ~VulkanPixelHistoryDiscardedFragmentsCallback()
+  {
+    for(const VkPipeline &pipe : m_PipesToDestroy)
+      m_pDriver->vkDestroyPipeline(m_pDriver->GetDev(), pipe, NULL);
+  }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return;
+
+    const rdcarray<int32_t> primIds = m_Events[eid];
+
+    VulkanRenderState prevState = m_pDriver->GetCmdRenderState();
+    VulkanRenderState &state = m_pDriver->GetCmdRenderState();
+    const VulkanCreationInfo::Pipeline &p =
+        m_pDriver->GetDebugManager()->GetPipelineInfo(state.graphics.pipeline);
+    // Create a pipeline with a scissor and colorWriteMask = 0, and disable all tests.
+    VkPipeline newPipe =
+        CreatePipeline(state.graphics.pipeline, eid, p.dynamicStates[VkDynamicScissor]);
+    state.graphics.pipeline = GetResID(newPipe);
+    state.BindPipeline(m_pDriver, cmd, VulkanRenderState::PipelineBinding::BindGraphics, false);
+    for(uint32_t i = 0; i < primIds.size(); i++)
+    {
+      uint32_t queryId = (uint32_t)m_OcclusionIndices.size();
+      ObjDisp(cmd)->CmdBeginQuery(Unwrap(cmd), m_OcclusionPool, queryId, 0);
+      const DrawcallDescription *drawcall = m_pDriver->GetDrawcall(eid);
+      uint32_t primId = primIds[i];
+      // TODO once pixel history distinguishes between instances, draw only the instance for
+      // this fragment
+      if(drawcall->flags & DrawFlags::Indexed)
+        ObjDisp(cmd)->CmdDrawIndexed(
+            Unwrap(cmd), RENDERDOC_NumVerticesPerPrimitive(drawcall->topology),
+            RDCMAX(1U, drawcall->numInstances),
+            drawcall->indexOffset + RENDERDOC_VertexOffset(drawcall->topology, primId),
+            drawcall->baseVertex, drawcall->instanceOffset);
+      else
+        ObjDisp(cmd)->CmdDraw(
+            Unwrap(cmd), RENDERDOC_NumVerticesPerPrimitive(drawcall->topology),
+            RDCMAX(1U, drawcall->numInstances),
+            drawcall->vertexOffset + RENDERDOC_VertexOffset(drawcall->topology, primId),
+            drawcall->instanceOffset);
+      ObjDisp(cmd)->CmdEndQuery(Unwrap(cmd), m_OcclusionPool, queryId);
+
+      m_OcclusionIndices[make_rdcpair<uint32_t, uint32_t>(eid, primId)] = queryId;
+    }
+    m_pDriver->GetCmdRenderState() = prevState;
+    m_pDriver->GetCmdRenderState().BindPipeline(
+        m_pDriver, cmd, VulkanRenderState::PipelineBinding::BindGraphics, false);
+  }
+
+  VkPipeline CreatePipeline(ResourceId pipe, uint32_t eid, bool dynamicScissor)
```

Comment:
Not needed for this PR, but in future - do you think it would be possible and worthwhile to share some of this logic between callbacks? either with static helper functions in this file or by deriving from a common parent class?

Reply:
In the new commit I added common methods to the parent class of all callbacks:
- one to create a pipeline create info that can be further tweaked
- one to create renderpass with DS attachment, and optionally replacing the format of the colour attachment
- one to create framebuffer with DS attachmnet, and optionally replacing the color attachment view